src/ringmesh/geomodel/tools/geomodel_tools.cpp

```diff
@@ -0,0 +1,165 @@
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#include <array>
+#include <iomanip>
+#include <iostream>
+
+#include <geogram/basic/command_line.h>
+#include <geogram/basic/progress.h>
+
+#include <ringmesh/basic/geometry.h>
+#include <ringmesh/geomodel/core/geomodel.h>
+#include <ringmesh/geomodel/core/geomodel_entity.h>
+#include <ringmesh/geomodel/core/geomodel_geological_entity.h>
+#include <ringmesh/geomodel/core/geomodel_mesh_entity.h>
+#include <ringmesh/geomodel/tools/geomodel_tools.h>
+#include <ringmesh/mesh/mesh.h>
+
+#include <ringmesh/tetrahedralize/tetra_gen.h>
+
+/*!
+ * @file Set of high level API functions
+ */
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    void geomodel_tools_api copy_geomodel(
+        const GeoModel< DIMENSION >& from, GeoModel< DIMENSION >& to )
+    {
+        GeoModelBuilder< DIMENSION > to_builder( to );
+        to_builder.topology.copy_topology( from );
+        to_builder.geometry.copy_meshes( from );
+        to_builder.geology.copy_geology( from );
+    }
+
+    template < index_t DIMENSION >
+    void translate( GeoModel< DIMENSION >& geomodel,
+        const vecn< DIMENSION >& translation_vector )
+    {
+        GeoModelBuilder< DIMENSION > builder( geomodel );
+        for( auto v : range( geomodel.mesh.vertices.nb() ) )
+        {
+            // Coordinates are not directly modified to
+            // update the matching vertices in geomodel entities
+            const auto& p = geomodel.mesh.vertices.vertex( v );
+            builder.geometry.set_mesh_entity_vertex( v, p + translation_vector );
+        }
+    }
+
+    void rotate( GeoModel3D& geomodel,
+        const vec3& origin,
+        const vec3& axis,
+        double angle,
+        bool degrees )
+    {
+        if( length( axis ) < geomodel.epsilon() )
+        {
+            Logger::err( "GeoModel",
+                "Rotation around an epsilon length axis is impossible" );
+            return;
+        }
+        GEO::Matrix< 4, double > rot_mat{ rotation_matrix_about_arbitrary_axis(
+            origin, axis, angle, degrees ) };
+
+        GeoModelBuilder3D builder( geomodel );
+        for( auto v : range( geomodel.mesh.vertices.nb() ) )
+        {
+            const vec3& p = geomodel.mesh.vertices.vertex( v );
+            std::array< double, 4 > old{ { p[0], p[1], p[2], 1. } };
+            std::array< double, 4 > new_p{ { 0, 0, 0, 1. } };
+            GEO::mult( rot_mat, old.data(), new_p.data() );
+            ringmesh_assert( std::fabs( new_p[3] - 1. ) < global_epsilon );
+
+            builder.geometry.set_mesh_entity_vertex( v, vec3{ new_p.data() } );
+        }
+    }
+
+    void tetrahedralize(
+        GeoModel3D& geomodel, index_t region_id, bool add_steiner_points )
+    {
+        /* @todo Review: Maybe rethink these functions
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS1135)